### PR TITLE
Jmon/printf

### PR DESCRIPTION
**Pull request by johanmon**
How does this work for adding dynamic width and precision fields? 

Two boolean fields are added to the Spect struct: dynamic_width and dynamic_precision. The parsing of the format string will detect '*' in either position. 

The number of arguments now needed for a Format specification needs to  take the dynamic arguments into account. 

Computing the required buffer length becomes more complicated since we also need to check dynamic arguments. 

The formatting now includes a pre-stage where the dynamic flags are checked, if present the arguments are used to fill in the right values of a Spec specifications. 

Completely new to Julia so might be all wrong but it simulates functionality ok :-)

### CODE REVIEW

File: base/printf.jl
Comment (c42f) [inline base/printf.jl]: It would be better to avoid this if possible. Having the spec be mutable can have performance consequences and is generally just less clear than working with immutable state.

**Review by c42f — CHANGES_REQUESTED** [review]
Hi and welcome! I think this standard printf functionality is reasonable feature to have (given that we have printf at all), and the code I saw at a quick glance looks pretty natural.

It's worth adding a cross reference to the PR which this is based on: https://github.com/JuliaLang/julia/pull/32859 which will need to be merged first.

Based on a very quick first look at the code here's some things to change:
* I suspect you can avoid making the struct mutable and should do so if possible.
* There has been some recent code churn and Printf has now moved to `stdlib/Printf`. Unfortunately you'll need to rebase your branch at some stage to fix the conflicts. It may be worth waiting until #32859 has been rebased.
* Some tests are required to cover all the cases where `*` may be used in the various combinations. These should go in stdlib/Printf/test/runtests.jl (after the rebase)

Happy first pull request!

File: base/printf.jl
Comment (johanmon) [reply base/printf.jl]: Will fix. 

We can avoid having the Spec mutable, the change will then check the dynamic property in each fmt method - not a problem. 

Tests - will add. 

Nota bene - this is my first git pl in a large collaborative project so I'm also learning the in and outs of git.  Things like "adding a cross ref" is still greek to me and I'm not from Greece but will figure out :-)

Comment (c42f) [reply base/printf.jl]: Sorry to be obscure! All I mean is that it's worth adding links (cross references) in your pull request description where relevant so that people can understand the context (not talking about git here). Links like this:

https://github.com/JuliaLang/julia/pull/32859

Comment (johanmon) [reply base/printf.jl]: got it :-) 

[#32859](https://github.com/JuliaLang/julia/pull/32859)

File: base/printf.jl
Comment (c42f) [inline base/printf.jl]: This sets `argp = +1`.  I assume you meant `argp += 1` to increment `argp`?

File: base/printf.jl
Comment (c42f) [inline base/printf.jl]: There's some trailing space here (and elsewhere) to remove.

File: base/printf.jl
Comment (c42f) [reply base/printf.jl]: (Add some tests which catch this bug :-) )

**Review by c42f — COMMENTED** [review]
Thanks, I think this is a bit blocked by #32859 for the moment, though in the meantime I think you could add more tests. In particular handling of multiple arguments with variable width and precision.

Other than that, please rebase once #32859 is merged and we can have a closer look here.

File: base/printf.jl
Comment (c42f) [reply base/printf.jl]: (see something like `git diff origin/jq/printf...jmon/printf` to show the whitespace)

File: stdlib/Printf/test/runtests.jl
Comment (c42f) [inline stdlib/Printf/test/runtests.jl]: Better, I think, to write these tests very concretely.  For example,

```julia
@test (Printf.@printf("%*d", 10, 12)) == "        12"
```

This way it's super clear what the expected result is because it's right there in the test case.

Comment (c42f) [reply stdlib/Printf/test/runtests.jl]: Note also that `@printf` always returns `nothing` and prints to stdout as a side effect.

So these tests will always pass!  (you need `@sprintf`)

Comment (johanmon) [reply stdlib/Printf/test/runtests.jl]: stupid me :-)

**Comment by johanmon** [comment]
rewrote this since everything move and I was changing things anyway, the new pull request is in: 

https://github.com/JuliaLang/julia/pull/40105

**Pull request closed without merging**
